### ISSUE TITLE
Allowing converters to get called without arguments.

### ISSUE THREAD

**Issue by Tails0630** [Rapptz/discord.py#6383]
**Summary**

Commands.Convert using Optinal[str] argument rather than mandatory

**What is the feature request for?**

discord.ext.commands

**The Problem**

I'm trying to implement a converter that has different behavior depending on the argument and even if it isn\'t given.

**The Ideal Solution**

It'd be quite useful if the converters work even if no argument is provided, taking into account that they currently don't get executed if no argument is passed, rather than allowing the convert function to handle it by itself in such cases.

**The Current Solution**

The current solution is basically using Optional str, in the command itself, and only refer to Converter if an argument is provided, if it's not provided, then using conditionals within the core to handle the function when no parameters are passed.

```py
from typing import Optional
from discord.ext import commands
class Reference(commands.Converter):
    async def convert(self, ctx: commands.context.Context, argument: Optional[str]):
        if not argument: return 0
        if argument.isdigit(): return 1
        return 2

@bot.command()
async def value(ctx, *, num: Reference):
    await ctx.send(num)
```

**Comment by Rapptz** [comment]
This isn't gonna happen.

If you want a default parameter you're free to provide one yourself:

```py
@bot.command()
async def value(ctx, *, num: Reference = 1):
    await ctx.send(num)
```

**Comment by Tails0630** [comment]
Ah, I mostly suggested because I was trying to implement a default parameter that uses the context's information, in order to retrieve a message, which goes this way but the argument requirement causes the function to not get executed.

```py
class Reference(commands.Converter):
    async def convert(self, ctx: commands.context.Context, argument: Optional[str]):
        def process(message: Optional[discord.Message], ctx: commands.context.Context):
            if not message: return None
            if message.author == ctx.bot.user and message.embeds:
                return message
            return None
        try:
            if not argument:
                ref_msg : discord.MessageReference = ctx.message.reference
                channel: discord.TextChannel = ctx.channel
                message: discord.Message = await channel.fetch_message(ref_msg.message_id)
                return await process(message, ctx)
            else:
                URL = REGEX_URL.match(argument)
                if URL:
                    guild_id, channel_id, message_id = map(int, URL.groups())
                    guild: discord.Guild = ctx.bot.get_guild(guild_id)
                    channel: discord.TextChannel = guild.get_channel(channel_id)
                    message: discord.Message = await channel.fetch_message(message_id)
                    return await process(message, ctx)
                elif argument.isdigit():
                    channel: discord.TextChannel = ctx.channel
                    message: discord.Message = await channel.fetch_message(int(argument))
                    return await process(message, ctx)
                return None
        except Exception:
            return None
```

**Comment by Rapptz** [comment]
The answer is mostly the same. A different approach is implemented in #1849
